sphinx/util/docutils.py

```diff
@@ -273,6 +274,26 @@ def switch_source_input(state, content):
         state.memo.reporter.get_source_and_line = get_source_and_line
 
 
+class SphinxDirective(Directive):
+    """A base class for Directives.
+
+    Compared with ``docutils.parsers.rst.Directive``, this class improves
+    accessibility to Sphinx APIs.
+    """
+
+    @property
+    def env(self):
+        # type: () -> BuildEnvironment
+        """Reference to the :class:`.BuildEnvironment` object."""
+        return self.state.document.settings.env
+
+    @property
+    def config(self):
+        # type: () -> Config
+        """Reference to the :class:`.Config` object."""
+        return self.env.config
+
+
```

Comment:
Would it be worth exposing `app` too?

    self.state.document.settings.env.app

Reply:
Good point. The reason why I did not add `app` is making Sphinx components less coupled with `app` object. I agree with `app` is very useful because it behaves as a hub. With using it, we can access all of the component.
But I guess what we really want to get is not `app` object. There are very few features belongs to `app` object.
In addition, on testing, providing `app` object is a bit expensive. To create it, we have to initialize all of submodules; config, extensions, env, builders and so on.

But this is just my idea, and not designed well. I know many extensions are strongly coupled with `app`. And it is very important and useful at present. So I can add it if it is reasonable.